### ISSUE TITLE
[Pending - Rajat Payment] [$1000] Add payment method button jumps when tapping back button on Paypal page

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14843]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to settings> payments
2. Click on Add payment method button and choose paypal option
3. Click on back button

## Expected Result:
Add payment method button should not jump

## Actual Result:
Add payment method button jumps

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] iOS / native

**Version Number:**  v1.2.65-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/216862183-2acbfd52-5063-47f0-be91-b2f04088d8c5.MP4

https://user-images.githubusercontent.com/43996225/216862205-679fac15-51a2-4bd2-8e55-d0bf338f77a2.MP4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @gadhiyamanan 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675419194137479

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01d9d6800730a96ae0</li>
        <li>Upwork Job ID: 1622954994768916480</li>
        <li>Last Price Increase: 2023-02-07</li>
    </ul>
</details>

**Comment by NicMendonca** [comment]
https://user-images.githubusercontent.com/24947345/217261967-040c25dc-c6ec-4bbc-af37-af3aca8a57a8.MP4

Can confirm its *only* happening with the `paypal` option

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01d9d6800730a96ae0

**Comment by MelvinBot** [comment]
Current assignee @NicMendonca is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @mananjadhav (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mountiny (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by tienifr** [comment]
# Proposal

## Problem
When we navigate from the PayPal screen (which at the time has keyboard open) back to the Payment page, the keyboard closing behavior interferes with the Payment page View's layout, causing the `Add payment method` button to jump.

## Solution
We should disable the `KeyboardAvoidingView` behavior in the Payment page to prevent this. The Payment page does not have any text input, the keyboard never actually opens here so it will not cause any issue.

```diff
diff --git a/src/components/ScreenWrapper/index.js b/src/components/ScreenWrapper/index.js
index e7ba7e5d0e..64ddd46319 100644
--- a/src/components/ScreenWrapper/index.js
+++ b/src/components/ScreenWrapper/index.js
@@ -96,7 +96,11 @@ class ScreenWrapper extends React.Component {
                                 paddingStyle,
                             ]}
                         >
-                            <KeyboardAvoidingView style={[styles.w100, styles.h100, {maxHeight: this.props.windowHeight}]} behavior={this.props.keyboardAvoidingViewBehavior}>
+                            <KeyboardAvoidingView
+                                style={[styles.w100, styles.h100, {maxHeight: this.props.windowHeight}]}
+                                behavior={this.props.keyboardAvoidingViewBehavior}
+                                enabled={this.props.keyboardAvoidingViewEnabled}
+                            >
                                 <HeaderGap />
                                 {// If props.children is a function, call it to provide the insets to the children.
                                     _.isFunction(this.props.children)
diff --git a/src/components/ScreenWrapper/propTypes.js b/src/components/ScreenWrapper/propTypes.js
index 46cc9ed631..08aaaa151c 100644
--- a/src/components/ScreenWrapper/propTypes.js
+++ b/src/components/ScreenWrapper/propTypes.js
@@ -28,6 +28,9 @@ const propTypes = {
         /** Indicates when an Alert modal is about to be visible */
         willAlertModalBecomeVisible: PropTypes.bool,
     }),
+
+    /** The enabled flag to pass to the KeyboardAvoidingView */
+    keyboardAvoidingViewEnabled: PropTypes.bool,
 };
 
 const defaultProps = {
@@ -37,6 +40,7 @@ const defaultProps = {
     onTransitionEnd: () => {},
     modal: {},
     keyboardAvoidingViewBehavior: 'padding',
+    keyboardAvoidingViewEnabled: true,
 };
 
 export {propTypes, defaultProps};
diff --git a/src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js b/src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js
index 9a480454ea..5f8c0a6212 100644
--- a/src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js
+++ b/src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js
@@ -342,7 +342,7 @@ class BasePaymentsPage extends React.Component {
         // Determines whether or not the modal popup is mounted from the bottom of the screen instead of the side mount on Web or Desktop screens
         const isPopoverBottomMount = this.state.anchorPositionTop === 0 || this.props.isSmallScreenWidth;
         return (
-            <ScreenWrapper>
+            <ScreenWrapper keyboardAvoidingViewEnabled={false}>
                 <HeaderWithCloseButton
                     title={this.props.translate('common.payments')}
                     shouldShowBackButton

```

**Comment by tienifr** [comment]
Working well after the fix:

https://user-images.githubusercontent.com/113963320/217284088-4daf5280-2bf3-4130-a5f3-365472beafb5.mov

**Comment by tienifr** [comment]
# Proposal 2

I managed to find the actual root cause of this. This happens due to the `focusPayPalMeInput` is called both when we navigate into and out of the PayPal screen (the `onTransitionEnd` are triggered in both cases). In order to fix this we need to check that we're not going back to previous screen before calling the `focusPayPalMeInput`.

```diff
diff --git a/src/pages/settings/Payments/AddPayPalMePage.js b/src/pages/settings/Payments/AddPayPalMePage.js
index ba2b5ec994..2b3aa88ae1 100644
--- a/src/pages/settings/Payments/AddPayPalMePage.js
+++ b/src/pages/settings/Payments/AddPayPalMePage.js
@@ -23,6 +23,7 @@ class AddPayPalMePage extends React.Component {
         this.state = {
             payPalMeUsername: '',
             payPalMeUsernameError: false,
+            didGoBack: false,
         };
         this.setPayPalMeUsername = this.setPayPalMeUsername.bind(this);
         this.focusPayPalMeInput = this.focusPayPalMeInput.bind(this);
@@ -54,11 +55,18 @@ class AddPayPalMePage extends React.Component {
 
     render() {
         return (
-            <ScreenWrapper onTransitionEnd={this.focusPayPalMeInput}>
+            <ScreenWrapper onTransitionEnd={() => {
+                if (!this.state.didGoBack) {
+                    this.focusPayPalMeInput();
+                }
+            }}>
                 <HeaderWithCloseButton
                     title={this.props.translate('common.payPalMe')}
                     shouldShowBackButton
-                    onBackButtonPress={() => Navigation.navigate(ROUTES.SETTINGS_PAYMENTS)}
+                    onBackButtonPress={() => {
+                        this.setState({didGoBack: true});
+                        Navigation.navigate(ROUTES.SETTINGS_PAYMENTS);
+                    }}
                     onCloseButtonPress={() => Navigation.dismissModal(true)}
                 />
                 <View style={[styles.flex1, styles.p5]}>

```

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by tienifr** [comment]
> calling focusPayPalMeInput inside InteractionManager.runAfterInteractions inside componentDidMount instead (this seems to be the cleanest solution)

Here's the diff for variation 1 of the above solution https://github.com/Expensify/App/compare/main...tienifr:App:fork/fix/14843-proposal-3?expand=1, it's clean and works well

**Comment by tienifr** [comment]
We might also want to take a look at other places where we use `onTransitionEnd` to make sure the same issue doesn't happen. It's out of scope for this issue though IMO but I'm happy to take a look upon request.

**Comment by s77rt** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
On back navigation the add payment method button jumps

### What is the root cause of that problem?
On back navigation after hiding the screen we are focusing the paypal input that causes the keyboard to briefly shows up resulting in the jump behaviour. The focus call is initialised by the `ScreenWrapper`'s callback `onTransitionEnd`.

### What changes do you think we should make in order to solve the problem?
Do not focus the paypal input `onTransitionEnd`. Instead pass `autoFocus` and `shouldDelayFocus` to the `<TextInput />`.

_We can apply the same fix to other effected pages_

### What alternative solutions did you explore? (Optional)
None

**Comment by tienifr** [comment]
@s77rt I thought about using `autoFocus` and `shouldDelayFocus` too but that causes a minor UI glitch for me (the Button of the next screen is pushed up by the keyboard before the Button of the previous screen fully disappears).

Also I think `shouldDelayFocus` is a hack that uses `setTimeout` which should only be used for Android when we have no other options. Like in [here](https://github.com/Expensify/App/pull/12513/files), all of its usage needs to have comments explaining

**Comment by s77rt** [comment]
Thanks for the feedback @tienifr. I do have answers for your concerns but let's not focus on the technical details for now not to create unnecessary back and forth communications.

**Comment by tienifr** [comment]
### Updated proposal

Another variation from [here](https://github.com/Expensify/App/issues/14843#issuecomment-1420961600) is to use `autoFocus`, but with `InteractionManager.runAfterInteractions` added to the `BaseTextInput` as in [this diff](https://github.com/Expensify/App/compare/main...tienifr:App:fork/fix/14843-proposal-2-autofocus?expand=1)

**Comment by MelvinBot** [comment]
📣 @parasharrajat You have been assigned to this job by @mountiny!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by s77rt** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/14843#issuecomment-1421543948)

**Comment by mountiny** [comment]
Thanks everyone for the proposals.

I agree with the feedback to @tienifr by Rajat, it would be great to show one finished proposal instead of multiple comments. @s77rt I really appreciate you attitude in this issue and the write-up you made in the proposal is excellent and in line with what the new proposals should look like, so great job on that!

The proposals are all close to me, however, since I have to choose one I prefer the last one from @parasharrajat, I think it nice and clean to use the `e.data.closing` properties react-navigation provides us and the final code would be quite easy to understand.

I appreciate this one was close and I can see other proposal owner might feel like they should have got this job, but I am sure you will land next job since your talent is clearly there. Thanks for collaboration here!

@parasharrajat please feel free to raise a PR for this issue.

**Comment by tienifr** [comment]
Thanks @parasharrajat and @mountiny for the feedback!

The proposal template was recently introduced so I wasn't aware of it yet. Sorry for any inconvenience.

**Comment by mountiny** [comment]
No worries, thank YOU!

**Comment by s77rt** [comment]
@mountiny Thanks for the quick review. I'm okay going with @parasharrajat 's proposal but wouldn't it be better to have that logic in `<TextInput />` instead? (we would still use `e.data.closing`, in my proposal I referred to that by "**if we are opening the page** focus the input")

**Comment by mananjadhav** [comment]
@mountiny Thanks for moving this forward, but now that @parasharrajat has submitted a proposal, do you want me to be the C+ on this one?

I am inclined towards implementing the `e.data.closing` to `ScreenWrapper`, which should cover all the pages.

**Comment by mountiny** [comment]
@mananjadhav you can still be here as a reviewer and tested. I like a suggestion of moving this as high as possible so it covers all possible cases. But we will have to be even more careful with testing all the other scenarios.

@s77rt I feel like both would work just fine and there is note many arguments for one or another, at least I cannot find them easily. But I think screenWrapper might make more sense since this is related to transitions lifecycle

**Comment by mountiny** [comment]
waiting for the pr

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.73-1 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15050


If no regressions arise, payment will be issued on 2023-02-24. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@mananjadhav / @parasharrajat / @mountiny] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@mananjadhav / @parasharrajat / @mountiny] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@mananjadhav / @parasharrajat / @mountiny] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@mananjadhav / @parasharrajat] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@NicMendonca] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by mountiny** [comment]
@parasharrajat @mananjadhav  Would you be bale to handle the checklist points please? Thanks!

**Comment by mananjadhav** [comment]
I'll try to update the checklist by today/tomorrow

**Comment by MelvinBot** [comment]
📣 @mananjadhav! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by mananjadhav** [comment]
@mountiny @parasharrajat 

**Regression Test Proposal**

1. Open the app.
2. Go to Settings page.
3. Open the Payments Page
4. Click on Add Payments Button -> Choose Paypal.me
5. Check that input is focused by default on opening the page.
6. Press Arrow icon from the top left, to go back.
6. Notice that the `Add Payments Button` page doesn't flicker/jump. 


Do we agree 👍 or 👎

----

Apart from that I tried to track the PR but I think this existed long before. @parasharrajat do you have any insights on which PR must've caused this regression?

**Comment by mountiny** [comment]
looks good, I think we are fine without the PR

**Comment by mananjadhav** [comment]
@NicMendonca this is ready for payout today. Can you please help with that?

**Comment by mountiny** [comment]
@NicMendonca Should be able to get to this today 🙌

**Comment by mananjadhav** [comment]
@mountiny @NicMendonca Quick bump on this one.

**Comment by NicMendonca** [comment]
I am so sorry @mananjadhav @parasharrajat can you please apply to the job/ accept my offer: https://www.upwork.com/jobs/~01d9d6800730a96ae0

**Comment by gadhiyamanan** [comment]
@NicMendonca applied for reporting

**Comment by NicMendonca** [comment]
@gadhiyamanan yep - I got yours!

**Comment by NicMendonca** [comment]
@gadhiyamanan paid - thank you!

**Comment by gadhiyamanan** [comment]
@NicMendonca please close the upwork contract 🙇‍♂️

**Comment by NicMendonca** [comment]
@mananjadhav paid!

**Comment by NicMendonca** [comment]
@parasharrajat I will check back here for your payment tomorrow [because of this](https://expensify.slack.com/archives/C02NK2DQWUX/p1677613613743399). Thanks!

**Comment by NicMendonca** [comment]
Everyone had been paid 🎉  I am incredibly sorry again for the delay! Thank you!
